docs/architecture/adr-070-hasher.md

```diff
@@ -0,0 +1,146 @@
+# ADR 070: hasher 
+
+## Changelog
+
+- 26 July 2021: Initial Draft (@jayt106)
+
+## Status
+
+Proposed
+
+## Context
+
+The current hashing algorithm in the Tendermint core project relies on the Go standard package [sha256](https://pkg.go.dev/crypto/sha256) and the implementation [crypto/tmhash](https://github.com/tendermint/tendermint/blob/master/crypto/tmhash/hash.go) only. The user might want to extend the hash algorithm for the different project's
+requirements. For example, the [ethermint](https://github.com/tharsis/ethermint) tries to bring the transaction to the Ethereum network but the transaction hash in the Ethereum world represents it in a different way [RLP hash](https://github.com/ethereum/go-ethereum/blob/92b8f28df3255c6cef9605063850d77b46146763/core/types/transaction.go#L368).
+
+Alternatively, some advanced hash algorithm provides faster-hashing speed, it might increase some scabilities to the project, like the mempool processing, trie proccessing, database data store/retrieve, and so on. For example, the supplementary Go cryptography libraries implemented [blake2b](https://pkg.go.dev/golang.org/x/crypto/blake2b) has roughly 2.25X faster than the `sha256` we mentioned above (both testing under the AVX2 supported CPU processor) [ref](https://github.com/SimonWaldherr/golang-benchmarks#hash). The [sha256-simd](https://github.com/minio/sha256-simd/) provides up to 8x faster than the `crypto/sha256` if the CPU processor supports [AVX512](https://en.wikipedia.org/wiki/AVX-512). The [xxHash](https://github.com/cespare/xxhash) has been using in [Polkadot](https://substrate.dev/docs/en/knowledgebase/advanced/cryptography#hashing-algorithms) project and it provides dramatically speed hashing can be using certain scenarios like the hash of immutable data input.
```

Comment:
IMO, it would be good to also link to existing discussions around these topics: 
- https://github.com/tendermint/tendermint/issues/5631
- https://github.com/tendermint/tendermint/issues/2186

Reply:
Thanks for the feedback, I will add it into the ADR.